detectors/node/opentelemetry-resource-detector-instana/src/detectors/InstanaAgentDetector.ts

```diff
@@ -0,0 +1,145 @@
+
+  async detect(_config?: ResourceDetectionConfig): Promise<Resource> {
+    const host =
+      process.env.INSTANA_AGENT_HOST || this.INSTANA_AGENT_DEFAULT_HOST;
+    const port = Number(
+      process.env.INSTANA_AGENT_PORT || this.INSTANA_AGENT_DEFAULT_PORT
+    );
+
+    const data = await this._retryHandler(host, port, 0);
+
+    return new Resource({
+      [SemanticResourceAttributes.PROCESS_PID]: data.pid,
+      'instana.host.id': data.agentUuid,
+      // NOTE: The agent reads this property to correlate the process with the service
+      'service.instance.id': data.pid.toString(),
+    });
+  }
+
+  private timeout(ms: number) {
+    return new Promise(resolve => setTimeout(resolve, ms));
+  }
+
+  private async _retryHandler(
+    host: string,
+    port: number,
+    tries: number
+  ): Promise<any> {
+    const MAX_TRIES = 3;
+
+    try {
+      return await this._fetchAgentData(host, port);
+    } catch (err) {
+      if (tries < MAX_TRIES) {
+        console.log(
+          `Retrying to connect to the Instana agent on ${host}:${port}....`
+        );
+
+        const retryTimeout = process.env.INSTANA_RETRY_TIMEOUT
+          ? Number(process.env.INSTANA_RETRY_TIMEOUT)
+          : 1000;
+
+        await this.timeout(retryTimeout);
+        return await this._retryHandler(host, port, tries + 1);
+      }
+
+      throw err;
+    }
+  }
+
+  private async _fetchAgentData(host: string, port: number): Promise<any> {
+    const agentTimeout = process.env.INSTANA_AGENT_TIMEOUT
+      ? Number(process.env.INSTANA_AGENT_TIMEOUT)
+      : 3 * 1000;
+
+    const payload = {
+      pid: process.pid,
+    };
+
+    const payloadStr = JSON.stringify(payload);
+    const contentLength = Buffer.from(payloadStr, 'utf8').length;
+
+    return new Promise((resolve, reject) => {
+      // agent needs more than 1s!
+      const timeoutId = setTimeout(() => {
+        req.destroy();
+        reject(new Error('Instana Agent request timed out.'));
+      }, agentTimeout);
+
+      const req = http.request(
+        {
+          host,
+          port,
+          method: 'PUT',
+          path: '/com.instana.plugin.nodejs.discovery',
+          headers: {
+            Accept: 'application/json',
+            'Content-Type': 'application/json; charset=UTF-8',
+            'Content-Length': contentLength,
+          },
+        },
+        res => {
+          clearTimeout(timeoutId);
+          res.setEncoding('utf8');
+          let rawData = '';
+
+          res.on('data', chunk => (rawData += chunk));
+          res.on('end', () => {
+            if (
+              res.statusCode &&
+              res.statusCode >= 200 &&
+              res.statusCode < 300
+            ) {
+              try {
+                resolve(JSON.parse(rawData));
+              } catch (e) {
+                reject(e);
+              }
+            } else {
+              return reject(
+                new Error('Failed to load page, status code: ' + res.statusCode)
```

Comment:
What do you mean by "load page"? I guess the error message can be more informative

Reply:
This error message is used in all detectors. So I kept the pattern.